pkg/kv/kvserver/reports/critical_localities_report.go

```diff
@@ -387,71 +390,70 @@ func (v *criticalLocalitiesVisitor) countRange(
 	// Any of the localities of any of the nodes could be critical. We'll check
 	// them one by one.
 	for _, loc := range dedupLocal {
-		processLocalityForRange(ctx, r, zoneKey, v.report, loc, v.nodeChecker, stores)
+		processLocalityForRange(ctx, r, zoneKey, loc, v.nodeChecker, stores, v.report)
 	}
 }
 
-// processLocalityForRange checks a single locality constraint against a
-// range with replicas in each of the stores given, contributing to rep.
+// processLocalityForRange checks whether a given locality loc is critical for
+// range r with replicas in each of the stores given. If the locality is found
+// to be critical, rep will be updated. A locality is critical if the range
+// cannot make progress if all the replicas in the locality were to become
+// unavailable (in addition to the replicas indicated by the nodeChecker to also
+// be unavailable).
+//
+// storeDescs are the descriptors for the stores of all the replicas, in order.
+// Empty descriptors are passed in for stores about which we're missing info;
+// such stores will not be considered to be in loc.
+//
+// Note that if a range is unavailable to begin with, the localities of all its
+// replicas are considered to be critical.
 func processLocalityForRange(
 	ctx context.Context,
 	r *roachpb.RangeDescriptor,
 	zoneKey ZoneKey,
-	rep LocalityReport,
 	loc roachpb.Locality,
 	nodeChecker nodeChecker,
 	storeDescs []roachpb.StoreDescriptor,
+	rep LocalityReport,
 ) {
-	// Compute the required quorum and the number of live nodes. If the number of
-	// live nodes gets lower than the required quorum then the range is already
-	// unavailable.
-	quorumCount := len(r.Replicas().Voters())/2 + 1
-	liveNodeCount := len(storeDescs)
-	for _, storeDesc := range storeDescs {
-		isStoreLive := nodeChecker(storeDesc.Node.NodeID)
-		if !isStoreLive {
-			if liveNodeCount >= quorumCount {
-				liveNodeCount--
-				if liveNodeCount < quorumCount {
-					break
-				}
+	// inLoc returns whether other is the same, or a sub-locality of loc.
+	inLoc := func(other roachpb.Locality) bool {
+		// If other is shorter, then it's either unrelated to loc, or more generic
+		// than loc. The "more generic" case shouldn't happen with nodes with
+		// properly-configured localities, since other comes from a node's locality
```

Comment:
Hmm. This isn't exactly an invariant here, right? Can't we start a node with locality `foo` and another one with `foo,bar`? Shouldn't we be able to handle this as well as we can, even if it's a silly thing to do?